File: src/main/java/com/example/cqrs/command/entity/event/AccountCreatedEventEntity.java
package com.example.cqrs.command.entity.event;

import com.example.cqrs.command.entity.event.metadata.EventMetadata;
import jakarta.persistence.DiscriminatorValue;
import jakarta.persistence.Entity;
import lombok.AccessLevel;
import lombok.NoArgsConstructor;

import java.time.LocalDateTime;

@NoArgsConstructor(access = AccessLevel.PROTECTED)
@DiscriminatorValue("AccountCreatedEvent")
@Entity
public class AccountCreatedEventEntity extends AbstractAccountEventEntity {

    public AccountCreatedEventEntity(String accountId, LocalDateTime eventDate,
                                     double amount, EventMetadata metadata) {
        super(accountId, eventDate, amount, metadata);
    }

    // factory method
    public static AccountCreatedEventEntity of(String accountId, LocalDateTime eventDate, double amount, EventMetadata metadata) {
        return new AccountCreatedEventEntity(accountId, eventDate, amount, metadata);
    }

}

File: src/main/java/com/example/cqrs/command/entity/event/AccountCreatedEventEntity.kt
package com.example.cqrs.command.entity.event

import com.example.cqrs.command.entity.event.event.AbstractAccountEventEntity
import com.example.cqrs.command.entity.event.metadata.EventMetadata
import jakarta.persistence.DiscriminatorValue
import jakarta.persistence.Entity
import java.time.LocalDateTime

@Entity
@DiscriminatorValue("AccountCreatedEvent")
class AccountCreatedEventEntity(
    accountId: String,
    eventDate: LocalDateTime,
    amount: Double,
    metadata: EventMetadata
) : AbstractAccountEventEntity(
    id = null,
    accountId = accountId,
    eventDate = eventDate,
    amount = amount,
    metadata = metadata
) {
    companion object {
        fun of(
            accountId: String,
            eventDate: LocalDateTime,
            amount: Double,
            metadata: EventMetadata
        ): AccountCreatedEventEntity {
            return AccountCreatedEventEntity(
                accountId = accountId,
                eventDate = eventDate,
                amount = amount,
                metadata = metadata
            )
        }
    }
}